mlir/lib/Linker/IRMover.cpp

```diff
@@ -7,618 +7,46 @@
-  if (auto f = dyn_cast<Function>(src))
-    return linkFunctionBody(cast<Function>(dst), f);
-  if (auto gv = dyn_cast<GlobalVariable>(src.getOperation()))
-    return linkGlobalVariable(cast<GlobalVariable>(dst), gv);
-  if (auto ga = dyn_cast<GlobalAlias>(src.getOperation()))
-    return linkAliasAliasee(cast<GlobalAlias>(dst), ga);
-  return linkIFuncResolver(cast<GlobalIFunc>(dst), cast<GlobalIFunc>(src));
-}
-
-Expected<Operation *> MLIRLinker::linkAppendingVarProto(GlobalVariable dst,
-                                                        GlobalVariable src) {
-  llvm_unreachable("Not implemented");
-}
-
-void MLIRLinker::flushRAUWWorklist() {
-  for (const auto &[oldOp, newOp] : rauwWorklist) {
-    if (auto sym = dyn_cast<SymbolOpInterface>(newOp)) {
-      auto name = sym.getNameAttr();
-      if (failed(SymbolTable::replaceAllSymbolUses(oldOp, name, composite)))
-        oldOp->emitError("unable to replace all symbol uses for ") << name;
-    }
-    oldOp->erase();
-  }
-  rauwWorklist.clear();
-}
+LogicalResult IRMover::move(ArrayRef<ConflictPair> valuesToLink) {
+  worklist.insert(worklist.end(), valuesToLink.rbegin(), valuesToLink.rend());
 
-Error MLIRLinker::run() {
-  // TODO: is metadata materialization needed? Is that even a thing here?
-
-  // TODO: Do we need to care about dbginfoformat, data layout and target
-  // triple?
-
-  // TODO: Do we need to care about CUDA?
-
-  // TODO: Compute type mapping, if needed?
-
-  std::reverse(worklist.begin(), worklist.end());
   while (!worklist.empty()) {
-    GlobalValue gv = worklist.back();
+    ConflictPair pair = worklist.back();
     worklist.pop_back();
 
-    LLVM_DEBUG(llvm::dbgs() << "Linking: " << gv.getLinkedName() << "\n");
-
-    Operation *gvo = gv.getOperation();
-    if (valueMap.contains(gvo) || indirectSymbolValueMap.contains(gvo))
-      continue;
+    assert(!mapping.contains(pair.src) && "expected no mapping for source");
 
-    assert(!gv.isDeclaration());
-
-    mapper.mapValue(gv);
-
-    if (err)
-      return std::move(*err);
-    flushRAUWWorklist();
+    if (!remap(pair))
+      return failure();
   }
 
-  doneLinkingBodies = true;
-  mapper.addFlags(RF_NullMapMissingGlobalValues);
+  return success();
+}
 
-  // Reorder the globals just added to the destination module to match their
-  // original order in the source module.
-  src->walk([&](GlobalVariable gv) {
-    if (gv.hasAppendingLinkage())
-      return WalkResult::skip();
-    if (auto newValue = mapper.mapValue(gv)) {
-      if (auto newGv = dyn_cast<GlobalVariable>(newValue)) {
-        newGv->remove();
-        composite->getRegion(0).back().push_back(newGv);
-      }
-    }
+Operation *IRMover::remap(ConflictPair pair) {
+  Operation *remaped = materialize(pair);
+  if (!remaped)
+    return nullptr;
 
-    // do not recurse into global values
-    return WalkResult::skip();
-  });
+  // TODO this should be part of liker interface
+  if (!composite->isProperAncestor(remaped)) {
```

Comment:
Why do we need this? Isn't it an error to materialize into something that isn't `composite`?